presto-spi/src/main/java/com/facebook/presto/spi/block/StructBuilder.java

```diff
@@ -0,0 +1,226 @@
+
+        return output.slice();
+    }
+
+    public StructBuilder addAll(Object... values)
+    {
+        for (Object value : values) {
+            add(value);
+        }
+
+        return this;
+    }
+
+    public StructBuilder addAll(List<?> values)
+    {
+        values.forEach(this::add);
+
+        return this;
+    }
+
+    public StructBuilder add(Block block, int position)
+    {
+        if (block.isNull(position)) {
+            builder.appendNull();
+        }
+        else {
+            block.writeBytesTo(position, 0, block.getLength(position), builder);
+            builder.closeEntry();
+        }
+        return this;
+    }
+
+    public StructBuilder addAll(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            add(block, i);
+        }
+
+        return this;
+    }
+
+    public StructBuilder add(Object value)
+    {
+        if (!typeIterator.hasNext()) {
+            throw new IllegalStateException("Cannot add element on a unknown type");
+        }
+        Type type = typeIterator.next();
+        if (value == null) {
+            builder.appendNull();
+            return this;
+        }
+        if (type.getJavaType() == boolean.class) {
+            type.writeBoolean(builder, (boolean) value);
+        }
+        else if (type.getJavaType() == long.class) {
+            type.writeLong(builder, ((Number) value).longValue());
+        }
+        else if (type.getJavaType() == double.class) {
+            type.writeDouble(builder, ((Number) value).doubleValue());
+        }
+        else if (type.getJavaType() == Slice.class) {
+            type.writeSlice(builder, toSlice(type, value));
+        }
+        else {
+            throw new UnsupportedOperationException("Unsupported type " + type);
+        }
+
+        return this;
+    }
+
+    private Slice toSlice(Type type, Object value)
+    {
+        Slice slice;
+        if (value instanceof String) {
+            slice = Slices.utf8Slice((String) value);
+        }
+        else if (value instanceof Slice) {
+            slice = (Slice) value;
+        }
+        else if (value instanceof List) {
+            StructBuilder structBuilder = new StructBuilder(type.getTypeParameters());
+            ((List<?>) value).forEach(structBuilder::add);
+            slice = structBuilder.build();
+        }
+        else if (value instanceof Map) {
+            StructBuilder structBuilder = new StructBuilder(type.getTypeParameters());
+            ((Map<?, ?>) value).forEach((k, v) -> structBuilder.add(k).add(v));
+
+            slice = structBuilder.build();
+        }
+        else if (value instanceof byte[]) {
+            slice = Slices.wrappedBuffer((byte[]) value);
+        }
+        else {
+            throw new UnsupportedOperationException("Unsupported type " + value.getClass());
+        }
+
+        return slice;
+    }
+
+    private class CycleIterator implements Iterator<Type>
```

Comment:
I think this class is overkill. I'd just embed this logic in the enclosing class.